pkg/kubernetes/kubectl_apply.go

```diff
@@ -0,0 +1,207 @@
+	"k8s.io/client-go/discovery/cached/disk"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/kubectl/pkg/cmd/apply"
+	kubectlutil "k8s.io/kubectl/pkg/util"
+	"k8s.io/kubectl/pkg/util/openapi"
+	"k8s.io/kubectl/pkg/validation"
+)
+
+type Applier struct {
+	restConfig *rest.Config
+	discovery  discovery.CachedDiscoveryInterface
+	kubeClient kubernetes.Interface
+	maxRetries uint64
+}
+
+func NewApplier(restConfig *rest.Config, maxRetries uint64) (*Applier, error) {
+	discoveryCacheDir, err := os.MkdirTemp("", "discovery-cache")
+	if err != nil {
+		return nil, err
+	}
+	httpCacheDir, err := os.MkdirTemp("", "http-cache")
+	if err != nil {
+		return nil, err
+	}
+
+	discoveryInterface, err := disk.NewCachedDiscoveryClientForConfig(restConfig, discoveryCacheDir, httpCacheDir, 5*time.Minute)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Applier{
+		restConfig: restConfig,
+		discovery:  discoveryInterface,
+		kubeClient: client,
+		maxRetries: maxRetries,
+	}, nil
+}
+
+func (d *Applier) Apply(ctx context.Context, manifest *ResourceManifest) error {
+	return d.applyKubernetesManifest(ctx, manifest)
+}
+
+func (d *Applier) applyKubernetesManifest(ctx context.Context, manifest *ResourceManifest) error {
+	validator := validation.ConjunctiveSchema{
+		validation.NewSchemaValidation(d),
+		validation.NoDoubleKeySchema{},
+	}
+
+	obj, gvk, err := decodeAndValidate(manifest, validator)
+	if err != nil {
+		return err
+	}
+
+	mapping, err := d.getRESTMappingFor(gvk)
+	if err != nil {
+		return err
+	}
+
+	client, err := d.getRestClientFor(gvk)
+	if err != nil {
+		return err
+	}
+
+	applyManifest := func() error {
+		return d.applyKubernetesResource(ctx, manifest, obj, client, mapping)
+	}
+	return backoff.Retry(applyManifest, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), d.maxRetries))
+}
+
+func (d *Applier) applyKubernetesResource(_ context.Context, manifest *ResourceManifest, obj runtime.Object, client *rest.RESTClient, mapping *meta.RESTMapping) error {
+	helper := resource.NewHelper(client, mapping).WithFieldManager(apply.FieldManagerClientSideApply)
+
+	// Get the modified configuration of the object. Embed the result
+	// as an annotation in the modified configuration, so that it will appear
+	// in the patch sent to the server.
+	modified, err := kubectlutil.GetModifiedConfiguration(obj, true, unstructured.UnstructuredJSONScheme)
+	if err != nil {
+		return err
+	}
+
+	current, err := helper.Get(manifest.Namespace, manifest.Name)
+	if err != nil {
+		if !apiErrors.IsNotFound(err) {
+			return err
+		}
+
+		// Create the resource if it doesn't exist
+		// First, update the annotation used by kubectl apply
+		if err := kubectlutil.CreateApplyAnnotation(obj, unstructured.UnstructuredJSONScheme); err != nil {
+			return err
+		}
+
+		// TODO: created log
```

Comment:
Do we want to resolve this? This looks more like a DEBUG log level entry, if we support a debug mode for CLM 🤔